lighthouse-cli/test/smokehouse/run-smoke.js

```diff
@@ -200,6 +198,7 @@ async function cli() {
   if (failingTests.length && (process.env.RETRY_SMOKES || process.env.CI)) {
     console.log('Retrying failed tests...');
     for (const failedResult of failingTests) {
+      /** @type {number} */
```

Comment:
This was weird. Without the explicit type annotation, this error is emitted:

```
'resultIndex' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.
```

Reply:
that is weird... @brendankenny? :)

Reply:
Some more clues for @brendankenny. I think it's likely a TypeScript bug.

Error is also avoided with this explicit cast:

```
/** @type {{
    id: string;
    error?: Error | undefined;
  }[]} */
  const smokeResults = [];
```

the troublesome line is this index setter:

```
smokeResults[resultIndex] = (await runSmokehouse([smokeDefn]))[0];
```

Without that line, there's no error. When I hover over `smokeResults` in `const resultIndex = smokeResults.indexOf(failedResult);`, the type is correct, but hover over it in `smokeResults[resultIndex] = (await runSmokehouse([smokeDefn]))[0];` and it is `any[]`.

The type inferencing is all out of whack here. `smokeResults` is inferred to be 
```
{
    id: string;
    error?: Error | undefined;
}[]
```
but only after the `for..of..runSmokehouse` loop. `resultIndex` is correctly given `number`, but the TSC gets confused when attempting to use that derived type from the original inferred variable on that same variable whose type was inferred (that makes sense. yeah? yea).

Reply:
Minimal example:

![image](https://user-images.githubusercontent.com/4071474/50525011-d7e0a100-0a8d-11e9-8bd5-72180d39054b.png)

Reply:
yes, this is a typescript limitation. `smokeResults` is being inferred as an evolving array type, adding element types as the type checker reads our code.

`arr[n] = someValue` is one of the evolving initializer patterns that adds `typeof someValue` to the array type if `n` is numeric. If the type checker is being semi-smart the circularity is based on the fact that when checking that `resultIndex` is numeric, it finds it's from `Array<T>.indexOf`, where the array type `T` is (in part) based on the assignment below to `smokeResults[resultIndex]`. A special case would probably need to be added to note that some array methods always return a number, regardless of the type of the array.

If the type checker is being less smart, it's probably just running into a control flow limitation, where the `indexOf` query and the assignment are preventing reduction of the inferred type across the loop with the assignment above.

Reply:
Also, good minimal example, you should file it (though it may be another variant of https://github.com/Microsoft/TypeScript/issues/19955).

Not sure if it just wasn't the point of what was being demonstrated, but `blah.ts` in that screenshot has the same issue if you turn on `noImplicitAny` ([playground link](http://www.typescriptlang.org/play/#src=const%20arr%20%3D%20%5B%5D%3B%0A%0Aarr.push(1)%3B%0Aarr%3B%0A%0Awhile%20(true)%20%7B%0A%20%20%20%20const%20index%20%3D%20arr.indexOf(1)%3B%0A%20%20%20%20arr%5Bindex%5D%20%3D%201%3B%0A%7D) (have to turn on the flag in options))

Reply:
oh yeah, end of all this is to work around it, the index needs an explicit type (as you've done here), the array needs an explicit type, or don't save an intermediate variable so there's no need for a circular definition (`smokeResults[smokeResults.indexOf(failedResult)] = (await runSmokehouse([smokeDefn]))[0];`). What you have is the shortest of the three, so seems good to me :)

Reply:
Great, I've filed an issue: https://github.com/Microsoft/TypeScript/issues/29290